threedi_schema/migrations/versions/0300_upgrade_db_settings.py

```diff
@@ -0,0 +1,296 @@
+
+    # set use_groundwater_flow
+    sql = """
+        UPDATE model_settings
+        SET use_groundwater_flow = CASE
+            WHEN (SELECT groundwater_hydraulic_conductivity FROM groundwater LIMIT 1) IS NOT NULL OR (SELECT groundwater_hydraulic_conductivity_file FROM groundwater LIMIT 1) IS NOT NULL THEN 1
+            ELSE 0
+        END;
+        """
+    op.execute(sql)
+
+
+def drop_columns(table, columns):
+    with op.batch_alter_table(table) as batch_op:
+        for column in columns:
+            batch_op.drop_column(column)
+
+
+def set_use_from_exists(table, settings_col):
+    op.execute(f"UPDATE model_settings SET {settings_col} = TRUE WHERE EXISTS (SELECT 1 FROM {table});")
+
+
+def delete_all_but_matching_id(table, settings_id):
+    op.execute(f"DELETE FROM {table} WHERE id NOT IN (SELECT {settings_id} FROM model_settings);")
+
+
+def delete_all_but_first_row(table):
+    op.execute(f"DELETE FROM {table} WHERE id NOT IN ("
+               f"SELECT id FROM "
+               f"(SELECT id FROM {table} ORDER BY id LIMIT 1) AS subquery);")
+
+
+def correct_dem_paths():
+    # remove path - this will not work with nested paths!
+    op.execute(f"UPDATE model_settings SET dem_file = SUBSTR(dem_file, INSTR(dem_file, '/') + 1)")
+
+
+def upgrade():
+    conn = op.get_bind()
+    # Only use first row of global settings
+    delete_all_but_first_row("v2_global_settings")
+    rename_tables(RENAME_TABLES)
+    # rename columns in renamed tables
+    for table_name, columns in RENAME_COLUMNS.items():
+        rename_columns(table_name, columns)
+    # make all columns in renamed tables, except id, nullable
+    for _, table_name in RENAME_TABLES:
+        make_all_columns_nullable(table_name)
+    # create new tables
+    create_new_tables(COPY_FROM_GLOBAL.keys())
+    # add empty columns to tables
+    add_columns_to_tables(ADD_COLUMNS)
+    # copy data from model_settings to new tables and columns
+    for dst_table, columns in COPY_FROM_GLOBAL.items():
+        move_multiple_values_to_empty_table("model_settings", dst_table, columns)
+    # copy data from model_settings to existing table
+    move_values_to_table('model_settings', 'flooding_threshold', 'numerical_settings', 'flooding_threshold')
+    # set several 'use' columns based on other settings
+    set_use_from_settings_id()
+    set_use_from_exists("interception", "use_interception")
+    # keep first row of settings tables that have no explicit mapping
+    delete_all_but_matching_id("numerical_settings", "numerical_settings_id")
+    # drop unused id columns in model_settings
+    unused_cols = [settings_id for _, settings_id, _ in GLOBAL_SETTINGS_ID_TO_BOOL] + ["numerical_settings_id"]
+    drop_columns('model_settings', unused_cols)
+    # remove relative path prefix from dem path
+    correct_dem_paths()
+
+
+def downgrade():
+    # Not implemented on purpose
+    print("Downgrade back from 0.3xx is not supported")
```

Comment:
This has to fail (otherwise the next migration will fail) - just `raise` here instead of `print`

Reply:
Create an exception in 2033e431ab01dfd9ee9ea71ed1252c61a38c7dde. I wanted to add this to the tests, but I don't see `downgrade` being used anywhere.

Reply:
That's ok